File: src/infrastructure/rest/contracts/baseline_standing.py
"""The wire form of how an artifact stands relative to the enterprise baseline.

`src/domain/baseline_standing.py` is the value; this is its contract. Two arms discriminated on
`kind`, so a client matches exhaustively on the same two cases the server does rather than inferring
them from which fields happen to be present.

**Non-optional wherever it is carried, and that is the whole point.** `is_global` is declared
`bool | None = None` two fields away, and two of its three hit serialisers omit it — type-legally,
because the contract permits absence. A reader cannot then tell "not global" from "nobody said". For
this value that distinction is the answer: absence would read as the enterprise baseline, which is
precisely the reading that makes a pending local change invisible.
"""

from __future__ import annotations

from typing import Annotated, Literal

from pydantic import Field

from src.domain.baseline_standing import BASELINE_KIND, PROPOSED_KIND
from src.infrastructure.rest.contracts.wire_shape import Closed


class EnterpriseBaselineStanding(Closed):
    """What is shown is exactly the enterprise baseline."""

    kind: Literal["enterprise-baseline"] = BASELINE_KIND


class ProposedStanding(Closed):
    """The baseline plus local changes that have not been accepted upstream."""

    kind: Literal["proposed"] = PROPOSED_KIND
    #: Every live change against this artifact, so a reader can open them.
    proposal_ids: list[str]
    #: Which parts differ — the question a boolean could not answer, and the reason this is a union.
    changed_fields: list[str]
    #: What the artifact was when the changes were written.
    base_revision: str
    #: Derived on read, never stored: whether the changes still apply to the baseline they name.
    condition: Literal["current", "stale", "conflicting"]


#: One artifact's standing on the wire. `Field(discriminator=...)` rather than a bare union so the
#: published schema is a `oneOf` with a mapping, and a generated client gets the same exhaustive
#: match the server has instead of a guess at which arm it received.
BaselineStandingContract = Annotated[
    EnterpriseBaselineStanding | ProposedStanding, Field(discriminator="kind")
]

File: tools/gui/src/domain/schemas/baselineStanding.ts
import { Schema } from 'effect'

/**
 * How an artifact stands relative to the enterprise baseline.
 *
 * An engagement repository reads enterprise artifacts through references and cannot write them, so a
 * change to one is *proposed* and lives locally until it is accepted upstream. Every artifact row the
 * backend serves says which of the two it is showing.
 *
 * Two arms discriminated on `kind`, matching the server's own closed union, so a view matches
 * exhaustively instead of inferring the case from which fields happen to be present.
 *
 * **Required, never optional.** `is_global` sits beside this on the same rows declared
 * `bool | None = None`, and two of its three server-side serialisers omit it — so a reader cannot
 * tell "not global" from "nobody said". Here that distinction *is* the answer: a missing value would
 * read as the baseline, showing a proposed artifact as accepted upstream.
 */
export const EnterpriseBaselineSchema = Schema.Struct({
  kind: Schema.Literal('enterprise-baseline'),
})

export const ProposedStandingSchema = Schema.Struct({
  kind: Schema.Literal('proposed'),
  /** Every live change against this artifact, so a reader can open them. */
  proposal_ids: Schema.Array(Schema.String),
  /** Which parts differ — the question a boolean could not answer. */
  changed_fields: Schema.Array(Schema.String),
  /** What the artifact was when the changes were written. */
  base_revision: Schema.String,
  /** Derived on read, never stored: whether the changes still apply to the baseline they name. */
  condition: Schema.Literal('current', 'stale', 'conflicting'),
  /**
   * Whether anyone upstream has been asked to look yet — `awaiting-review` as soon as *one* of
   * the changes has been sent, because that is when taking one back stops being private.
   */
  review: Schema.Literal('not-sent', 'awaiting-review'),
})

export const BaselineStandingSchema = Schema.Union(EnterpriseBaselineSchema, ProposedStandingSchema)

export type BaselineStanding = typeof BaselineStandingSchema.Type
export type ProposedStanding = typeof ProposedStandingSchema.Type

/** Whether this standing carries local changes — the one predicate most views need. */
export const isProposed = (standing: BaselineStanding): standing is ProposedStanding =>
  standing.kind === 'proposed'
